toolkit/tools/imagegen/configuration/packagerepo.go

```diff
@@ -0,0 +1,203 @@
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/network"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/safechroot"
+	"github.com/microsoft/CBL-Mariner/toolkit/tools/internal/shell"
+)
+
+// PackageRepo defines the RPM repo to pull packages from during the installation
+// or after the system is installed. The "Install" option indicates that the provided
+// repository configuration will be saved in the installed system if specified, and only
+// available during the installation process if not
+type PackageRepo struct {
+	Name    string `json:"Name"`
+	BaseUrl string `json:"BaseUrl"`
+	Install bool   `json:"Install"`
+}
+
+// UnmarshalJSON Unmarshals a PackageRepo entry
+func (p *PackageRepo) UnmarshalJSON(b []byte) (err error) {
+	// Use an intermediate type which will use the default JSON unmarshal implementation
+	type IntermediateTypePackageRepo PackageRepo
+	err = json.Unmarshal(b, (*IntermediateTypePackageRepo)(p))
+	if err != nil {
+		return fmt.Errorf("failed to parse [PackageRepo]: %w", err)
+	}
+
+	// Now validate the resulting unmarshaled object
+	err = p.IsValid()
+	if err != nil {
+		return fmt.Errorf("failed to parse [PackageRepo]: %w", err)
+	}
+	return
+}
+
+// IsValid returns an error if the PackageRepo struct is not valid
+func (p *PackageRepo) IsValid() (err error) {
+	err = p.NameIsValid()
+	if err != nil {
+		return
+	}
+
+	err = p.RepoUrlIsValid()
+	if err != nil {
+		return
+	}
+
+	return
+}
+
+// NameIsValid returns an error if the package repo name is empty
+func (p *PackageRepo) NameIsValid() (err error) {
+	if strings.TrimSpace(p.Name) == "" {
+		return fmt.Errorf("invalid value for package repo name (%s), name cannot be empty", p.Name)
+	}
+	return
+}
+
+// RepoUrlIsValid returns an error if the package url is invalid
+func (p *PackageRepo) RepoUrlIsValid() (err error) {
+	if strings.TrimSpace(p.BaseUrl) == "" {
+		return fmt.Errorf("invalid value for package repo URL (%s), URL cannot be empty", p.BaseUrl)
+	}
+
+	_, err = url.ParseRequestURI(p.BaseUrl)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func writeAdditionalFieldstoRepoFile(fileName string) (err error) {
+	const (
+		gpgKey    = "gpgkey=file:///etc/pki/rpm-gpg/MICROSOFT-RPM-GPG-KEY file:///etc/pki/rpm-gpg/MICROSOFT-METADATA-GPG-KEY\n"
+		enable    = "enabled=1\n"
+		gpgCheck  = "gpgcheck=0\n"
+		skip      = "skip_if_unavailable=True\n"
+		sslVerify = "sslverify=0\n"
+	)
+
+	additionalFields := gpgKey + enable + gpgCheck + skip + sslVerify
+	err = file.Append(additionalFields, fileName)
+	if err != nil {
+		logger.Log.Errorf("Failed to write additional fields")
+	}
+
+	return
+}
+
+func createCustomRepoFile(fileName string, packageRepo PackageRepo) (err error) {
+	err = file.Create(fileName, 0644)
+	if err != nil {
+		logger.Log.Errorf("Error creating file %s", fileName)
+		return
+	}
+
+	// Write the repo identifier field
+	repoId := "[" + packageRepo.Name + "]\n"
+	err = file.Append(repoId, fileName)
```

Comment:
In #3008 you first build the string and then append the file at the very end, which I think is a more elegant solution (with the addition of using `strings.Builder` instead of a raw string). I'd suggest switching to that approach. This will also keep our code more consistent.

Reply:
Modified the implementation